[DTensor] fix DTensorTestCase.destroy_pg() when device_type is "cpu" but CUDA device is available

XilunWu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #161015

**Summary**
When `device_id` is not None, barrier() will choose the accelerator of the most
pripority, which means if the test specifies to use CPU for testing while CUDA is
available on the host, the barrier() will use CUDA. To avoid this and better respect
`self.device_type`, we add this branch to enforce barrier() to use CPU when
`self.device_type` is CPU and other accelerator is also available.

**Test**
`pytest test/distributed/tensor/test_dtensor_testbase.py`

**Debugging Output**
```
# from init_process_group()
init pg: backend=gloo, device_id = None
default_pg has backend: gloo, device_types: [device(type='cuda'), device(type='cpu')]

# from barrier()
barrier: device_ids = [10], devices = [], device = None, PG=[device(type='cuda'), device(type='cpu')]
```

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta

XilunWu committed:
[do NOT land] PG gloo init debug

[ghstack-poisoned]

XilunWu committed:
Update on "[do NOT land] PG gloo init debug"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

kwen2501 commented:
```
default_pg has backend: gloo, device_types: [device(type='cuda'), device(type='cpu')]
```
This is expected, because Gloo supports both CUDA device and CPU device.

kwen2501 commented:
```
# from barrier()
barrier: device_ids = [10]
```
This indicates that the caller has passed `10` to `dist.barrier(...)`. I suggest searching where this is called.

XilunWu committed:
Update on "[do NOT land] PG gloo init debug"


**Test**
`pytest test/distributed/tensor/debug_tianyu_paste.py`

**Output**
```
# from init_process_group()
init pg: backend=gloo, device_id = None
default_pg has backend: gloo, device_types: [device(type='cuda'), device(type='cpu')]

# from barrier()
barrier: device_ids = [10], devices = [], device = None, PG=[device(type='cuda'), device(type='cpu')]
```

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

XilunWu commented:
The root cause of this error is attempt to create `torch.device('cuda:8')` on 8-GPU host. This is because `DTensorTestBase` assumes that `self.world_size <= 8` while this may not hold true in incoming customer demand. This PR includes a fix attempt but this could be risky because some ranks are now placing barrier()'s input tensor on the same device.  WDYT? cc @kwen2501

XilunWu committed:
Update on "[do NOT land] PG gloo init debug"


**Test**
`pytest test/distributed/tensor/debug_tianyu_paste.py`

**Output**
```
# from init_process_group()
init pg: backend=gloo, device_id = None
default_pg has backend: gloo, device_types: [device(type='cuda'), device(type='cpu')]

# from barrier()
barrier: device_ids = [10], devices = [], device = None, PG=[device(type='cuda'), device(type='cpu')]
```

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

XilunWu committed:
Update on "[do NOT land] PG gloo init debug"


**Test**
`pytest test/distributed/tensor/debug_tianyu_paste.py`

**Output**
```
# from init_process_group()
init pg: backend=gloo, device_id = None
default_pg has backend: gloo, device_types: [device(type='cuda'), device(type='cpu')]

# from barrier()
barrier: device_ids = [10], devices = [], device = None, PG=[device(type='cuda'), device(type='cpu')]
```

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

XilunWu committed:
Update on "[DTensor] fix DTensorTestCase.destroy_pg() when device_type is "cpu" but CUDA device is available"


**Summary**
When `device_id` is not None, barrier() will choose the accelerator of the most
pripority, which means if the test specifies to use CPU for testing while CUDA is
available on the host, the barrier() will use CUDA. To avoid this and better respect
`self.device_type`, we add this branch to enforce barrier() to use CPU when
`self.device_type` is CPU and other accelerator is also available.

**Test**
`pytest test/distributed/tensor/test_dtensor_testbase.py`

**Debugging Output**
```
# from init_process_group()
init pg: backend=gloo, device_id = None
default_pg has backend: gloo, device_types: [device(type='cuda'), device(type='cpu')]

# from barrier()
barrier: device_ids = [10], devices = [], device = None, PG=[device(type='cuda'), device(type='cpu')]
```

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta

[ghstack-poisoned]

XilunWu commented on torch/testing/_internal/distributed/_tensor/common_dtensor.py:
```diff
@@ -420,7 +423,18 @@ def destroy_pg(self, device_id: Optional[int] = None) -> None:
+            # To avoid this and better respect `self.device_type`, we add this branch to
+            # enforce barrier() to use CPU when `self.device_type` is CPU and other
+            # accelerator is also available.
+            dist.barrier()
```
> Does this work because of _get_object_coll_device(group) == "cpu" on this line?

Correct, this returns the object collective PG which is GLOO.

> could be made more clear, e.g. maybe by including some comment you made here.

let me think more about it. We can put it in another PR.

tianyu-l reviewed (approved):
LGTM, had a final question.

XilunWu commented:
@pytorchbot merge
